Resolver refactoring

oatkiller opened the pull request:
## Summary

Hey y'all. 
There's a few subjective changes here. Feel free to say 'I liked it that way.'

* In my last PR I removed all usages of a field in `DataState` called `relatedEventStats`. I forgot to remove the field itself. This PR removes the field.
* Move the calculation of the 'total' related events to the selector. NB: this is still calculated in the view because it was duplicated to begin with. I'll try to fix this up in my next PR.
* Replace `magFactorX` (the first element of the projection matrix) with `xScale` for consistency with naming elsewhere in resolver.
* Refactor calculation of `actionableButtonsTopOffset` for readability
* the middleware had a call to `dispatch` in the `try...catch` used for HTTP request control flow. If `dispatch` threw an error due to code in the reducer or middleware, the error would be treated the same as a HTTP error (e.g. returning 500 from the server.) Moved the dispatch out of try...catch so it can be handled in the same way as other runtime errors. NB: we should look to replace our standard try...catch approach with something better.
* The component `ProcessEventListNarrowedByType` was in a module called `panels/panel_content_related_list.tsx`. This moves it to `panels/process_event_list_narrowed_by_type` for readability

### Checklist

Just refatoring.

- [x] click around the app, stuff still seems to work

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Pinging @elastic/endpoint-data-visibility-team (Team:Endpoint Data Visibility)

elasticmachine commented:
Pinging @elastic/endpoint-app-team (Feature:Endpoint)

oatkiller commented on x-pack/plugins/security_solution/public/resolver/store/data/selectors.ts:
```diff
@@ -301,3 +301,48 @@ export function databaseDocumentIDToAbort(state: DataState): string | null {
+ * Return the 'total' stats (whatever that means) for a process by entity ID
+ * TODO, how does this work for legacy events?
+ */
+export const relatedEventTotalForProcess: (
```
What was the justification for this? I'd like to put it down in comments here.

oatkiller replied:
Found it in the other copy of the code: https://github.com/oatkiller/kibana/blob/resolver-refactoring/x-pack/plugins/security_solution/public/resolver/view/panels/panel_content_related_counts.tsx#L54

oatkiller replied:
Now that I understand what this is doing I see that this number doesn't match the number of events.  Is there any point in showing this?

jonathan-buttner replied:
@bkimmel you might want to comment here.

There is a difference between total events and the total amount of all the categories added up which is what is happening here. It really depends on what we want to communicate to the user. If we communicate that there are 5 total related events, when they look at the break down the values won't always added up to 5 because a single event could fall into multiple categories. So it could be confusing.

bkimmel replied:
For a set of related events with the following ECS categories:
`[DNS, File], [DNS, File], [Registry]`
We break counts in submenus as:
```
DNS: 2
File: 2
Registry: 1
```

There are 2 choices then, for how to display the total: `5` or `3`

If I'm not mistaken, I believe it was you, Robert, who suggested that `5` is less confusing (and I think everyone agreed). This is what `counts to 5`. There are implications in other places, but that's the crux of what's happening here.

oatkiller replied:
/shrug

oatkiller commented on x-pack/plugins/security_solution/public/resolver/store/data/selectors.ts:
```diff
@@ -301,3 +301,48 @@ export function databaseDocumentIDToAbort(state: DataState): string | null {
+        return null;
+      }
+      let total = 0;
+      for (const value of Object.values(stats.events.byCategory)) {
```
Is `events.byCategory` a structure we made up?

jonathan-buttner replied:
`events.byCategory` is what the backend returns. It's an object of number of events that fall in a particular category for example:

```
{
  'network': 3,
  'file': 2,
}
```

michaelolo24 commented on x-pack/plugins/security_solution/public/resolver/view/process_event_dot.tsx:
```diff
@@ -293,25 +295,14 @@ const ProcessEventDotComponents = React.memo(
-        break;
-    }
+    const actionableButtonsTopOffset =
+      (isShowingEventActions ? 3.5 : isShowingDescriptionText ? 1 : 21) * xScale + 5;
```
Nice, you don't have anything against nested ternary's from a readability standpoint?

oatkiller replied:
wanted your thoughts on this (esp. since you wrote the original.)

1. I found the original a lot to read due to boilerplate. Specifically, reading `actionableButtonsTopOffset` vs `actionableButtonsTopOffset` was hard.
2. The repeated part of `actionalButtonsBaseTopOffset + ` and `* magFactorX` required me to visually compared the 3 versions to see that they were the same.
3. part of my motivation to make this terse is because the function it is in is long, and has many variables that aren't related to each other. That isn't the best motivation to make something terse. We could move the logic out of the function instead.

michaelolo24 replied:
1. agreed, makes sense
2. also makes sense
3/4. I like to throw in the occasional switch every now and again 😂 
5. Even if it's terse, it's still readable, so no need to revert any of it. I like moving it into an external function though because it will be a better separation of concerns and eventually it can probably be removed altogether in favor of mostly pure css.

jonathan-buttner commented on x-pack/plugins/security_solution/public/resolver/store/data/selectors.ts:
```diff
@@ -301,3 +301,46 @@ export function databaseDocumentIDToAbort(state: DataState): string | null {
+      return () => null;
+    }
+    return (event: ResolverEvent) => {
+      const nodeStats = statsMap.get(uniquePidForProcess(event));
```
Just leaving the note here like we talked about before @oatkiller this function `uniquePidForProcess` will suffer from the same problem we talked about before where `unique_pid` will have collisions in the map just like the `entityId` function. We'll need to do the serialization or something for that function too.

oatkiller replied:
I'll get something formally added to the team's backlog about this.

kibanamachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57484/)
* Commit: 662cef40f0f8b84a36044cd4a17078a69654f4e1
* Flaky suites:
  * kibana-ciGroup1

---

### [Test Failures](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57484/testReport)

<details><summary>Kibana Pipeline / kibana-oss-agent / Plugin Functional Tests.test/plugin_functional/test_suites/core_plugins/logging·ts.core plugins plugin logging writes info_pattern context to custom pattern appender</summary>

[Link to Jenkins](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57484/testReport//junit/Plugin%20Functional%20Tests/test_plugin_functional_test_suites_core_plugins_logging·ts/Kibana_Pipeline___kibana_oss_agent___core_plugins_plugin_logging_writes_info_pattern_context_to_custom_pattern_appender)


#### Standard Out
```
Failed Tests Reporter:
  - Test has failed 2 times on tracked branches: https://github.com/elastic/kibana/issues/70149

[00:00:00]       │
[00:01:17]         └-: core plugins
[00:01:17]           └-> "before all" hook
[00:03:47]           └-: plugin logging
[00:03:47]             └-> "before all" hook
[00:03:47]             └-> does not write to custom appenders when not configured
[00:03:47]               └-> "before each" hook: global before each
[00:03:47]               └-> "before each" hook
[00:03:47]               │ proc [kibana]   log   [13:33:49.085] [info][core_logging][debug_json][plugins] i go to the default appender!
[00:03:47]               └- ✓ pass  (53ms) "core plugins plugin logging does not write to custom appenders when not configured"
[00:03:47]             └-> writes debug_json context to custom JSON appender
[00:03:47]               └-> "before each" hook: global before each
[00:03:47]               └-> "before each" hook
[00:03:47]               └- ✓ pass  (44ms) "core plugins plugin logging writes debug_json context to custom JSON appender"
[00:03:47]             └-> writes info_json context to custom JSON appender
[00:03:47]               └-> "before each" hook: global before each
[00:03:48]               └-> "before each" hook
[00:03:48]               └- ✓ pass  (42ms) "core plugins plugin logging writes info_json context to custom JSON appender"
[00:03:48]             └-> writes debug_pattern context to custom pattern appender
[00:03:48]               └-> "before each" hook: global before each
[00:03:48]               └-> "before each" hook
[00:03:48]               └- ✓ pass  (55ms) "core plugins plugin logging writes debug_pattern context to custom pattern appender"
[00:03:48]             └-> writes info_pattern context to custom pattern appender
[00:03:48]               └-> "before each" hook: global before each
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [dev_tools] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]                 │ debg browser[INFO] http://localhost:6121/9007199254740991/bundles/kbn-ui-shared-deps/kbn-ui-shared-deps.js 444:106112 "INFO: 2020-06-30T13:33:49Z
[00:03:48]                 │        Adding connection to http://localhost:6121/elasticsearch
[00:03:48]                 │
[00:03:48]                 │      "
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [management] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [rendering] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [custom-app-route] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [appleave1] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [appleave2] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [chromeless] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [foo] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]                 │ debg browser[WARNING] http://localhost:6121/9007199254740991/bundles/core_legacy_compat.bundle.js 2495:20 "App [bar] is using deprecated mount context. Use core.getStartServices() instead."
[00:03:48]               └-> "before each" hook
[00:03:48]               │ info Taking screenshot "/dev/shm/workspace/kibana/test/functional/screenshots/failure/core plugins plugin logging writes info_pattern context to custom pattern appender.png"
[00:03:49]               │ debg browser[DEBUG] http://localhost:6121/9007199254740991/bundles/plugin/usageCollection/usageCollection.plugin.js 950:16 "Reporting user-agent."
[00:03:49]               │ debg browser[DEBUG] http://localhost:6121/9007199254740991/bundles/plugin/usageCollection/usageCollection.plugin.js 950:16 "Reporting application changed to core_legacy_compat"
[00:03:49]               │ info Current URL is: http://localhost:6121/app/core_legacy_compat
[00:03:49]               │ info Saving page source to: /dev/shm/workspace/kibana/test/plugin_functional/failure_debug/html/core plugins plugin logging writes info_pattern context to custom pattern appender.html
[00:03:49]               └- ✖ fail: "core plugins plugin logging writes info_pattern context to custom pattern appender"
[00:03:49]               │

```

#### Stack Trace
```
{ Error: expected [ '' ] to sort of equal [ 'CUSTOM - PATTERN [plugins.core_logging.info_pattern][INFO ] log2' ]
    at Assertion.assert (packages/kbn-expect/expect.js:100:11)
    at Assertion.eql (packages/kbn-expect/expect.js:244:8)
    at Context.it (test/plugin_functional/test_suites/core_plugins/logging.ts:119:53)
  actual: '[\n  ""\n]',
  expected:
   '[\n  "CUSTOM - PATTERN [plugins.core_logging.info_pattern][INFO ] log2"\n]',
  showDiff: true }
```
</details>

---

### Build metrics
> ✅ unchanged

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"UNSTABLE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57484/","number":"57484","commit":"662cef40f0f8b84a36044cd4a17078a69654f4e1"}]}
PIPELINE-->

oatkiller commented on x-pack/plugins/security_solution/public/resolver/store/data/reducer.ts:
```diff
@@ -9,7 +9,6 @@ import { DataState } from '../../types';
 import { ResolverAction } from '../actions';
 
 const initialState: DataState = {
-  relatedEventsStats: new Map(),
```
The logic that used to populate this piece of state has been moved to the selector called `relatedEventsStats`

oatkiller replied:
See: https://github.com/elastic/kibana/pull/69728/files#diff-b013c0d85f67056fb127d2247c99a538L78

kqualters-elastic commented on x-pack/plugins/security_solution/public/resolver/view/panels/process_event_list_narrowed_by_type.tsx:
```diff
@@ -97,6 +97,7 @@ export const ProcessEventListNarrowedByType = memo(function ProcessEventListNarr
 }) {
   const processName = processEvent && event.eventName(processEvent);
   const processEntityId = event.entityId(processEvent);
+
```
I think more than just this file have a file name that doesn't line up with the component name. Would probably make sense to fix those as well

oatkiller replied:
I'm changing them as I find 'em. I don't think we need a rule that says '1:1 relationship between components and filenames.' That said, if a file has just one export, which is a component, it would be nice to name the component and the file the same.

kqualters-elastic commented on x-pack/plugins/security_solution/public/resolver/view/panel.tsx:
```diff
@@ -24,7 +24,7 @@ import { useResolverDispatch } from './use_resolver_dispatch';
 import { ResolverEvent, ResolverNodeStats } from '../../../common/endpoint/types';
 import { SideEffectContext } from './side_effect_context';
-import { ProcessEventListNarrowedByType } from './panels/panel_content_related_list';
+import { ProcessEventListNarrowedByType } from './panels/process_event_list_narrowed_by_type';
```
this file should probably be panels/index.tsx instead of just panel.tsx

bkimmel replied:
:red_circle: That's not a list of `process event`s - it's a list of related events like (registry creation, file creation, etc.). Also, if we're changing the name of one of the panel content components, we should change the prefix on all of them.

bkimmel replied:
I gave it the wrong name from the export -> The file name should stay the same, the export name should change,

bkimmel replied:
Maybe something like `relatedEventListNarrowedByType` ?

oatkiller replied:
I'll revert this part of the PR.

oatkiller commented:
> This ruptures a lot of things we're using in other places (counts, stats, etc.) I need a little time to understand this.

Per our discussion, I think I've addressed your concerns. Let me know if I can do anything else.

bkimmel commented on x-pack/plugins/security_solution/public/resolver/view/process_event_dot.tsx:
```diff
@@ -177,11 +179,11 @@ type EventDisplayName = typeof displayNameRecord[keyof typeof displayNameRecord]
   typeof unknownEventTypeMessage;
 
 /**
- * Take a gross `schemaName` and return a beautiful translated one.
```
:information_source: No fun allowed :straight_ruler:

oatkiller replied:
Just trying to get clarity.

bkimmel commented on x-pack/plugins/security_solution/public/resolver/view/process_event_dot.tsx:
```diff
@@ -4,14 +4,16 @@
  * you may not use this file except in compliance with the Elastic License.
```
:cool: Thank you for _not_ changing the name of process_event_dot . I have an attachment to the name. It's a nice reminder that everything starts as a dot.

bkimmel replied:
To @kqualters-elastic point, we could probably improve it in the future by making it _more_ like a dot and less like a giant thing with a bunch of code in it but :shrug:

oatkiller commented:
took out the part about changing file names. lets do it in a follow up w/ @bkimmel's help

oatkiller commented:
@bkimmel lets follow up re: anything 'off'. Otherwise I'll just keep doing more of it.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57658/)
* Commit: 54043fdbfdf7d056e1ddae66e5b7a73c9722e16e

### Build metrics
> ✅ unchanged

### History
* :yellow_heart: [Build #57484](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57484/) was flaky 662cef40f0f8b84a36044cd4a17078a69654f4e1

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"UNSTABLE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57484/","number":"57484","commit":"662cef40f0f8b84a36044cd4a17078a69654f4e1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/57658/","number":"57658","commit":"54043fdbfdf7d056e1ddae66e5b7a73c9722e16e"}]}
PIPELINE-->
